### PR TITLE
add base task

### PR DESCRIPTION
**Pull request by VladislavPatrushev**

### CODE REVIEW

File: src/tasks/base.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.__constructor = Constructor(hparams)
+        self._metric_manager = self.__constructor.configure_metrics_manager()
+        self._criterion = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.example_input_array = [torch.rand(1, *shape) for _, shape in self.__input_shapes.items()]
```
Comment (PososikTeam) [inline src/tasks/base.py:26]: Why not create it on to_onnx() method?

File: src/tasks/base.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+    def __init__(self, hparams: DictConfig):
+        """Init BaseTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.__constructor = Constructor(hparams)
+        self._metric_manager = self.__constructor.configure_metrics_manager()
```
Comment (PososikTeam) [inline src/tasks/base.py:22]: Where is self.save_hyperparameters(hparams)?

File: src/tasks/base.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.example_input_array = [torch.rand(1, *shape) for _, shape in self.__input_shapes.items()]
+
+    @abstractmethod
+    def forward(self, *args, **kwargs):
+        pass
+
+    @abstractmethod
+    def forward_with_gt(batch: dict) -> dict:
```
Comment (PososikTeam) [inline src/tasks/base.py:33]: i think need documentation

File: src/tasks/base.py (hunk 1 above)
Comment (PososikTeam) [inline src/tasks/base.py:26]: Why it's public, do you think user will overwrite it?

File: src/tasks/base.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        pass
+
+    def on_test_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def training_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the training epoch with the outputs of all training steps."""
+        avg_loss = torch.stack([x['loss'] for x in outputs]).mean()
+        self.log('train/loss', avg_loss, on_step=False, on_epoch=True)
```
Comment (PososikTeam) [inline src/tasks/base.py:59]: It must be overwrite, bcause JoitLoss return avg_loss, with tagged_loss_dict (key - name_loss, value - loss_value). It needed because we want log every loss separatly too.

File: src/tasks/base.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        self.log('step', self.current_epoch, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('train'))
+
+    def validation_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the validation epoch with the outputs of all validation steps."""
+        avg_loss = torch.stack(outputs).mean()
+        self.log('valid/loss', avg_loss, on_step=False, on_epoch=True)
```
Comment (PososikTeam) [inline src/tasks/base.py:66]: Must be overwrite as in training_epoch_end.

File: src/tasks/base.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        self.log('step', self.current_epoch, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('valid'))
+
+    def test_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of a test epoch with the output of all test steps."""
+        avg_loss = torch.stack(outputs).mean()
+        self.log('test/loss', avg_loss, on_step=False, on_epoch=True)
```
Comment (PososikTeam) [inline src/tasks/base.py:73]: Must be overwrite as in training_epoch_end.

File: src/tasks/base.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    def on_train_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def on_test_start(self) -> None:
+        # TODO check and load checkpoint
+        pass
+
+    def training_epoch_end(self, outputs: torch.tensor) -> None:
+        """Is calling at the end of the training epoch with the outputs of all training steps."""
```
Comment (PososikTeam) [inline src/tasks/base.py:57]: It? And on the follow documentation

File: src/tasks/base.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                        **onnx_params)
+
+    def on_train_end(self) -> None:
+        """Is calling at the end of training before logger experiment is closed."""
+        onnx_params = self._hparams.onnx_params
+        if onnx_params is not None:
+            self.to_onnx(onnx_params)
+
+    def configure_optimizers(self) -> Tuple[List, List]:
+        """Configure optimizers."""
```
Comment (PososikTeam) [inline src/tasks/base.py:88]: Documentation Returns

File: src/tasks/base.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                schedulers.append(None)
+        return optimizers, schedulers
+
+    def train_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or more PyTorch DataLoaders for training."""
+        data_loader = self.__constructor.create_dataloaders('train')
+        return data_loader
+
+    def val_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for prediction."""
```
Comment (PososikTeam) [inline src/tasks/base.py:107]: Why shuffle must be False? Do we really compute on only shuffle False? Also please add documentation

File: src/tasks/classification.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        """Init ClassificationTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__(hparams)
+
+        self.backbone = BACKBONES.get(self._hparams.backbone_name)(**self._hparams.backbone_params)
+
+        self._hparams.pooling_params['in_features'] = self.backbone.num_features
```
Comment (PososikTeam) [inline src/tasks/classification.py:24]: Now it will be function `self.backbone.get_output_channels()`

File: src/tasks/base.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        total_loss, tagged_loss_values = outputs
+        avg_loss = total_loss.mean()
+        self.log('train/avg_loss', avg_loss, on_step=False, on_epoch=True)
+
+        for tag, loss_value in tagged_loss_values.items():
+            self.log(f'train/{tag}', loss_value, on_step=False, on_epoch=True)
+
+        self.log('step', self.current_epoch, on_step=False, on_epoch=True)
+        self.log_dict(self._metric_manager.on_epoch_end('train'))
+
```
Comment (PososikTeam) [inline src/tasks/base.py:69]: Space line

File: src/tasks/base.py [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        """Abstract forward method for validation an test."""
+        pass
+
+    @abstractmethod
+    def forward_with_gt(batch: dict) -> dict:
+        """Abstract forward method for training(with ground truth labels)."""
+        pass
+
+    @abstractmethod
+    def training_step(self, *args, **kwargs):
```
Comment (VladVin) [inline src/tasks/base.py:40]: I think you don't need to add these abstract methods because they should be overrided in child classes directly from `pl.Module`

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/tasks/base.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+    def train_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or more PyTorch DataLoaders for training."""
+        data_loader = self.__constructor.create_dataloaders('train')
+        return data_loader
+
+    def val_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for prediction."""
+        phase = 'valid'
+        drop_last = self._hparams['data'][phase][0]['dataloader']['drop_last']
```
Comment (VladVin) [inline src/tasks/base.py:133]: Phase might be not present in configuration. Check config structure

File: src/tasks/base.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        """Hyperparameters that set in yaml file."""
+        return self._hparams
+
+    @property
+    def metric_manager(self):
+        """Metric manager."""
+        return self._metric_manager
+
+    @property
+    def criterion(self):
```
Comment (VladVin) [inline src/tasks/base.py:177]: Rename to `losses` as it's named in `Constructor`

File: src/tasks/base.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    def test_step_end(self, step_outputs: torch.tensor) -> torch.tensor:
+        return step_outputs.mean(dim=0, keepdim=True)
+
+    @property
+    def hparams(self):
+        """Hyperparameters that set in yaml file."""
+        return self._hparams
+
+    @property
+    def metric_manager(self):
```
Comment (VladVin) [inline src/tasks/base.py:172]: Rename to `metrics_manager`

File: src/tasks/base.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.save_hyperparameters(hparams)
+        self.__constructor = Constructor(hparams)
+        self._metric_manager = self.__constructor.configure_metrics_manager()
+        self._criterion = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.example_input_array = [torch.rand(1, *shape) for _, shape in self.__input_shapes.items()]
```
Comment (VladVin) [inline src/tasks/base.py:27]: How will it work if there are multiple inputs to the model?

File: src/tasks/base.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+
+    @abstractmethod
+    def forward(self, *args, **kwargs):
+        """Abstract forward method for validation an test."""
+        pass
+
+    @abstractmethod
+    def forward_with_gt(batch: dict) -> dict:
```
Comment (VladVin) [inline src/tasks/base.py:35]: Propbably I missed something, but where are `forward_infer`and `forward_infer_with_gt`?

**Review by VladVin — CHANGES_REQUESTED** [review]

File: src/tasks/base.py (hunk 17 above)
Comment (VladislavPatrushev) [reply src/tasks/base.py:40]: We agreed to add them in the next iterations.

File: src/tasks/classification.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        features = self.pooling(features)
+        prediction = self.head(features, target)
+        output = {'target': target, 'embeddings': features, 'prediction': prediction}
+        return output
+
+    def training_step(self, batch: dict) -> torch.tensor:
+        """Complete training loop."""
+        output = self.forward_with_gt(batch)
+        loss = self._losses(**output)
+        self._metrics_manager.update('train', **output)
```
Comment (PososikTeam) [inline src/tasks/classification.py:65]: `self._metrics_manager.update(Phase.TRAIN, **output)`?

File: src/tasks/classification.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        output = self.forward_with_gt(batch)
+        loss = self._losses(**output)
+        self._metrics_manager.update('valid', **output)
+        return loss
+
+    def test_step(self, batch: dict) -> torch.tensor:
+        """Complete test loop."""
+        output = self.forward_with_gt(batch)
+        loss = self._losses(**output)
+        self._metrics_manager.update('test', **output)
```
Comment (PososikTeam) [inline src/tasks/classification.py:79]: self._metrics_manager.update(Phase.TEST, **output)?

File: src/tasks/base.py [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+
+        if data_params is None:
+            return None
+
+        data_loader = self.__constructor.create_dataloaders('train')
+        return data_loader
+
+    def val_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for prediction."""
+        phase = 'valid'
```
Comment (PososikTeam) [inline src/tasks/base.py:131]: Enum?

File: src/tasks/base.py [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+            drop_last = data_param['dataloader']['drop_last']
+            if drop_last:
+                raise ValueError(f'DataLoader parametrs `drop_last` must be False in {phase} phase.')
+
+        data_loader = self.__constructor.create_dataloaders(phase)
+        return data_loader
+
+    def test_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for testing."""
+        phase = 'test'
```
Comment (PososikTeam) [inline src/tasks/base.py:148]: Enum?

File: src/tasks/base.py [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+            drop_last = data_param['dataloader']['drop_last']
+            if drop_last:
+                raise ValueError(f'DataLoader parametrs `drop_last` must be False in {phase} phase.')
+
+        data_loader = self.__constructor.create_dataloaders(phase)
+        return data_loader
+
+    def predict_dataloader(self) -> Optional[List[torch.DataLoader]]:
+        """Implement one or multiple PyTorch DataLoaders for prediction."""
+        phase = 'predict'
```
Comment (PososikTeam) [inline src/tasks/base.py:165]: Enum?

File: src/tasks/classification.py [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+
+import torch
+from omegaconf import DictConfig
+
+from src.registry import BACKBONES, HEADS, POOLINGS, TASKS
+from src.tasks.base import BaseTask
+
+
+@TASKS.register_class
+class ClassificationTask(BaseTask):
```
Comment (PososikTeam) [inline src/tasks/classification.py:11]: What's about predict method? It's would be test step?

File: src/tasks/base.py [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+        self.__constructor = Constructor(hparams)
+        self._metrics_manager = self.__constructor.configure_metrics_manager()
+        self._losses = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
+        self.__input_shapes = self._hparams.input_shapes
+        self.__input_dtypes = self._hparams.input_dtypes
+        self._input_tensors = []
+
+        for input_shape, input_dtype in zip(self.__input_shapes, self.__input_dtypes):
+            input_tensor = torch.rand(1, *input_shape).type(torch.__dict__[input_dtype])
```
Comment (VladVin) [inline src/tasks/base.py:32]:
```suggestion
            input_tensor = torch.rand(*input_shape).type(torch.__dict__[input_dtype])
```

File: src/tasks/base.py [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+from omegaconf import DictConfig
+
+from src.constructor.config_structure import Phase
+from src.constructor.constuctor import Constructor
+
+
+class BaseTask(LightningModule, ABC):
+    """An abstract class that represent main methods of tasks."""
+
+    def __init__(self, hparams: DictConfig):
```
Comment (PososikTeam) [inline src/tasks/base.py:15]: Add "TODO: change type to ConfigParams"

File: src/tasks/base.py [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+        """
+        super().__init__()
+        self.save_hyperparameters(hparams)
+        self.__constructor = Constructor(hparams)
+        self._metrics_manager = self.__constructor.configure_metrics_manager()
+        self._losses = self.__constructor.configure_losses()
+        self._hparams = self.__constructor.hparams
```
Comment (PososikTeam) [inline src/tasks/base.py:26]: Why not `self._hparams = hparams`?

File: src/tasks/base.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+        self.__input_shapes = self._hparams.input_shapes
```
Comment (PososikTeam) [inline src/tasks/base.py:27]: Look how it did on previous version it must be `self._hparams.task.inpu_shapes`
